### ISSUE TITLE
Nested promise is changing the parent promise

### ISSUE THREAD

**Issue by victorjussiani** [fluffynuts/synchronous-promise#7]
I love to use your lib cause it make my tests really simple, but I'm getting a behavior different from Promise implementation.

When I try to do something like this:

```
    searchA() {
        return SynchronousPromise.resolve('A');
    }

    searchB() {
        return SynchronousPromise.resolve('B');
    }

    searchC() {
        return SynchronousPromise.resolve('C');
    }

    let consultA = this.searchA();
    let consultB = this.searchB();
    let consultC = consultA.then(() => {
        return this.searchC();
    });

    SynchronousPromise.all([consultA, consultB, consultC])
        .then(([resultA, resultB, resultC]) => {
            console.log(resultA, resultB, resultC);
        });
```
It prints this:

`'C', 'B', undefined`

It seems that the nested promise is changing the parent promise, can you help me with this issue?

Thank you

**Comment by fluffynuts** [comment]
Hi

It looks like the code sample you've given is two parts of an ed6 class? The only way I can get this to be valid JS is by changing to:

```
class Stuff {
        searchA() {
          return SynchronousPromise.resolve("A");
        }

        searchB() {
          return SynchronousPromise.resolve("B");
        }

        searchC() {
          return SynchronousPromise.resolve("C");
        }

        doStuff() {
          let consultA = this.searchA();
          let consultB = this.searchB();
          let consultC = consultA.then(() => {
            var cResult = this.searchC();
            return cResult;
          });

          return SynchronousPromise.all([consultA, consultB, consultC])
        }
      }
```

And then I get the behavior you describe, when running the "then" off of the "doStuff" method.

Unfortunately, there's no quick-fix for this. SynchronousPromise does not return a new instance when you `.then()` -- instead, it maintains state internally and returns itself. So the first result is run to completion ("C" instead of "A"). I'll give it some thought, but it's very unlikely to be fixed in a hurry, I'm afraid, simply because of the nature of the existing code. 

You can still achieve a similar flow (having resolution of C delayed until A completes and getting all three values in the final handler) by having the `consultA.then` handler change:
```
let consultC = consultA.then(aResult => {
  return this.searchC(cResult => [aResult, cResult]);
});
```
Which results in your final handler getting
```
[ "A", "C" ], "B", null
```

I know it's not ideal and what I'm proposing is simply a work-around for this case. You could also make your test asynchronous for this case.
I'm only suggesting these workarounds because I can see that I won't have a quick fix for this case. I'd have to significantly change the logic within SynchronousPromise (which I'll look at, but I can't guarantee any quick turnaround on it, I'm afraid).

**Comment by fluffynuts** [comment]
I just want to update that I have been looking at this and haven't (as yet) found a neat solution. I've even considered re-writing (and it may get to that).

**Comment by fluffynuts** [comment]
This has, technically, been fixed with a complete ground-up rewrite. Well, rewrite of the _code_ not the _tests_ -- I maintained the test suite so that the original behavior could be guaranteed.

Now, I'd like to release a fixed package but I want to cause as little disruption as possible, so I'm trying to get input on versioning. Semver states that a change which isn't breaking should be a patch increment, one which adds without detracting should be a minor increment and a breaking change should be a major increment. 

The fact is that this code, according to the tests, exhibits all of the original behavior _except_ that referenced by this issue (so, not really a breaking change -- more of a bugfix). _However_, this is a full rewrite (ok, mostly full -- I just just re-include the original `all()` function, since it passes all testing and wasn't the source of the issue), I don't want to cause issues for downstream projects. So I'm considering making a major increment since that would (normally) require dependents to manually update to the next version, alleviating the pressure of an automatic upgrade causing headaches.

I'm not anticipating headaches -- I just don't want to be the source of them, hence the request for comment.

**Comment by jquense** [comment]
I'd appreciate a major bump just for caution, but I can live with whatever you decide :) if you you want a bit more code path exercising maybe we could try and run the yup tests against the new version and see if anything breaks? They aren't direct api tests but there are lot of them so we might uncover some behavior changes

**Comment by fluffynuts** [comment]
That's actually a great idea. I've just tried cloning yup, but `npm test` out-of-the-box (ie, just `git clone`, `npm install` and `npm test`) claims to have two failing tests (though the failure looks like a repeat of the same thing):

```
   Number types  notEqual  52 should be invalid

    AssertionError: expected true to deeply equal false
      
      at Function.assertEqual (node_modules/chai/lib/chai/core/assertions.js:485:19)
      at Function.ctx.(anonymous function) (node_modules/chai/lib/chai/utils/addMethod.js:41:25)
      at node_modules/chai-as-promised/lib/chai-as-promised.js:308:26
          at <anonymous>

 FAIL  test/number.js
   Number types  notEqual  52 should be invalid

    AssertionError: expected true to deeply equal false
      
      at Function.assertEqual (node_modules/chai/lib/chai/core/assertions.js:485:19)
      at Function.ctx.(anonymous function) (node_modules/chai/lib/chai/utils/addMethod.js:41:25)
      at node_modules/chai-as-promised/lib/chai-as-promised.js:308:26
          at <anonymous>
```

This is node 8.2.1, Linux.

What am I doing wrong here? Is there something I have to do in addition to `npm install`?

The more I've thought about it, the more I favor a major increment -- but I'd also really like the happy feeling I'd get from seeing a downstream project's test run properly with SP before releasing, as per your suggestion.

**Comment by jquense** [comment]
sorry I think those tests are just failing seperately. let me confirm and fix

**Comment by fluffynuts** [comment]
Well, the good news is that copying my SP into the node_modules of yup and re-running doesn't result in any other failures (:

**Comment by jquense** [comment]
ok master is updated. you should see 438 tests pass

**Comment by fluffynuts** [comment]
Ok, it's all working. I know a day may not be long enough for others to respond, but this issue has been open 9 days for a fix. In all truth, I started a rewrite at least 4 times and hit a block with one or another existing test -- so I started again, until I have what I have now.

So I'm going to release 2.0.0. Here goes (: Thanks for participating in the discussion and providing the idea to test using your framework (:
